### ISSUE TITLE
Turn on fuzzy matching for entities

### ISSUE THREAD

**Issue by marek-danel** [IBM/watson-assistant-workbench#48]
We need to turn fuzzy matching on manually after each deployment. We would like to be able to set an entity to be fuzzy matched in WAW or even all entities defined by enumeration of values and their synonyms could be set to be fuzzy matched by default as the generalization to unseen utterances is the main feature of WA. What do you think @tomasmacek @Tereza-Pytelova ?

**Comment by Tereza-Pytelova** [comment]
We can use the first line of entity file for this purpose, but then every entity file has to have special first line or we have to somehow specify that the first line has special meaning.
Or there can be a file with list of fuzzy matched entities.
What is your opinion @tomasmacek @ondrejtomas ?
Result should look like this one:
```
    {
      "entity": "entity_name",
      "values": [ ... ],
      "fuzzy_match": true
    },
```

**Comment by ondrejtomas** [comment]
I would prefer some kind of configuration file that will correspond to settings in UI - e.g. system entities, fuzzy...

**Comment by tomasmacek** [comment]
Discussing this s @Tereza-Pytelova . We agreed to keep the information in a single place (entity csv file).
Two files wold make the life difficult for the user who wold have to keep open both files)

**Comment by Tereza-Pytelova** [comment]
@ondrejtomas do you agree?

**Comment by ondrejtomas** [comment]
I always try to think what it means from portability perspective... when I would like to move some of my entities to the new project or even share some entities between different projects... I think that "fuzzy" is not property of the entity, but it is current setting of the WA... so I would still rather keep separated entities, intents... raw data... and those additional information keep in some kind of configuration... I would rather have it in any configuration file that we currently have (e.g. under section 'entity')... but it is 2 against 1... so I let it up to you. :)

**Comment by ondrejtomas** [comment]
Thank you to not oppress me, but this is not what I meant... I know that you can switch it on and off per entity... but it is not property of that list... the entity list should still remain raw... for different NLU (e.g. MAC) that can't do "fuzzy", those data still be usable... even you can imagine, that somebody would like to use WA with version where "fuzzy" was not implemented yet... my proposal was to create something like section 'entity' where you can specify for each entity if you want to set fuzzy on or off (there could be also default value 'on' to not specify it for all)... you can also imagine that you could completely turn on or off the whole entity in this section.

**Comment by tomasmacek** [comment]
I agree that there might be reasons to enforce override of the individual setting and there might be reasons to set a default. Those could be an optional parameters in config file of waw. 
I suggest to keep this functionality  under a separate issue and implement it when needed, to avoid lingering this issue?

**Comment by ondrejtomas** [comment]
Well, as I said before, it is up to you... but just for the record, I am against to modify any raw data like this... entity lists in this case.

**Comment by Tereza-Pytelova** [comment]
Postponing this to the next week due to non-trivial discussions.

**Comment by Tereza-Pytelova** [comment]
*Summarising the pros and cons of the two approaches mentioned above:*
### Use the first line of an entity file for fuzzy (and possibly other) settings

`name.csv`:
```
##fuzzy_match=true
Anne;Ane;Anna
Ben;
...
```

#### Cons
 - When using the same entity in different project with different fuzzy settings, entity file has to be copied and two (and possibly more) versions have to be maintained.

### Config file for entities

`entities.cfg`:
```
[entities]
name=not_fuzzy # entity name should not be fuzzy-matched
direction # entity direction has the default fuzzy setting
city # entity city has the default fuzzy setting

[entities_fuzzy]
fuzzy_default=true # turn on fuzzy match for all entities
```

This config file can be used for listing used entities (not all from entities directory will be taken to the resulting workspace)

#### Cons
 - Entity settings are separated from entity data which could be more difficult to maintain.

**Comment by tomasmacek** [comment]
ok, I think I can live the with second option :=)

**Comment by Tereza-Pytelova** [comment]
Due to non-trivial discussions postponing this to the next sprint.

**Comment by Tereza-Pytelova** [comment]
Due to vacations postponing this to the next sprint.

**Comment by Tereza-Pytelova** [comment]
Within replanning this was moved to the Sprint 28.

**Comment by balatjan** [comment]
There is lot of dust on this issue and here we are again at the point, where we could really use fuzzy matching 😸

**Comment by Tereza-Pytelova** [comment]
It will be better to have entity settings in `common.cfg`

**Comment by ZoningFraction** [comment]
I suggest the following approach:
```ini
[entities]
fuzzy=on # fuzzy matching is turned on for all entities (overrides the default OFF value)
entity_city.fuzzy=off # overrides the global setting for this particular entity
entity_country.fuzzy=off # same as above
# all other entities in the folder defined in common.entities use the global value
```
This would be similar to definition of CF sequences, which also use a prefix to identify them in the `[cloudfunctions]` section. Now this doesn't address the question of what entities we want to use (if we don't want to include all the entities in the folder), but I think that we can just create a folder that contains ONLY those entities we wanted. I don't really see a reason to configure that in the .cfg file.

**Comment by Tereza-Pytelova** [comment]
@tomasmacek what do you thing about Ales' proposal? I find it quite sufficient for fuzzy matching purpose.
